### PR TITLE
Perform sanity checks with Python3

### PR DESCRIPTION
**Pull request by yillkid**
Switch test suites invocation to Python3.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,6 +10,7 @@ Most use cases for micropayments involve a single user or device interacting
 repeatedly with a few vendors.
 
 ## Prerequisites
+`Iota-swarm-node` is compatible with Python 3.5.
```
Comment (jserv) [inline README.md]: The wording is misleading. Instead, python 3.5+ is required as [PyOTA](https://github.com/iotaledger/iota.lib.py) configures. You have to emphasize on the rationale in plain English descriptions, otherwise nobody can be aware of it.

File: mk/python.mk [hunk 2]
```diff
@@ -0,0 +1,12 @@
+PYTHON = python3
+PYTHON := $(shell which $(PYTHON))
+ifndef PYTHON
```
Comment (jserv) [inline mk/python.mk]: Add extra condition if `python` indicated by `$PATH` is exactly of version 3. Think of the case somebody built Python3 from source and the generated executable is `python`.

File: mk/python.mk [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+$(error "python3 is required.")
```
Comment (jserv) [inline mk/python.mk]: Since Python 3.5+ is required, you have to validate its version.

File: mk/python.mk [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+endif
+
+# check "iota" module in Python installation
+PY_CHECK_MOD_IOTA := $(shell $(PYTHON) -c "import iota" 2>/dev/null && \
+                       echo 1 || echo 0)
+ifneq ("$(PY_CHECK_MOD_IOTA)","1")
+    $(error "skip $@ because PyIOTA is not installed.")
```
Comment (jserv) [inline mk/python.mk]: It is `PyOTA` or `iota.lib.py`.

Comment (jserv) [reply mk/python.mk]: The error message is wrong because it does not "skip". Instead, we would take actions if both Python3 and PyOTA are absent. You shall dump informative diagnosis messages.

File: mk/python.mk [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+PYTHON := $(shell which $(PYTHON))
+ifndef PYTHON
+$(error "python3 is required.")
+endif
+
+# check "iota" module in Python installation
+PY_CHECK_MOD_IOTA := $(shell $(PYTHON) -c "import iota" 2>/dev/null && \
+                       echo 1 || echo 0)
+ifneq ("$(PY_CHECK_MOD_IOTA)","1")
+    $(error "skip $@ because PyOTA is not installed.")
```
Comment (jserv) [inline mk/python.mk]: Inform users of the diagnosis and information about trouble shooting. Change the wording from "skip" to something really making sense.

**Comment by jserv** [comment]
TODO: Validate Python installation from the perspectives of the following:
1. python3 package: Don't assume Debian/Ubuntu packages are available. Instead, you have to be aware of the case that somebody always build python3 from source (and you should try as well).
2. The version of python3: only 3.5 or later are known to work.
3. The availability of `PyOTA`: you have to mention and alert users if no package is installed.

File: README.md [hunk 6]
```diff
@@ -10,6 +10,7 @@ Most use cases for micropayments involve a single user or device interacting
 repeatedly with a few vendors.
 
 ## Prerequisites
+`iota-swarm-node`, writing with Python3+ compatible code as well as depending on [Pyota](https://github.com/iotaledger/iota.lib.py) is theoretically available for bot Python 3.5 and 3.6; yet, our testing result indicates that the environment for Python 3.6 is invalid in this case. Details as [Pyota issue#203](https://github.com/iotaledger/iota.lib.py/issues/203).
```
[the remark this answers is not in the document]
Comment (jserv) [reply README.md]: You might have to consider [virtualenv](https://virtualenv.pypa.io/) or [pythonbrew](https://github.com/utahta/pythonbrew).

[the remark this answers is not in the document]
Comment (jserv) [reply README.md]: "bot Python 3.5 and 3.6" ?

File: mk/python.mk [hunk 7]
```diff
@@ -0,0 +1,12 @@
+PYTHON = python3.5
```
[the remark this answers is not in the document]
Comment (jserv) [reply mk/python.mk]: The coverage about Python 3.7 is not addressed yet.

[the remark this answers is not in the document]
Comment (jserv) [reply mk/python.mk]: You shall always try to build Python from source and check how its directory structure is.

File: mk/python.mk [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+PYTHON := $(shell which $(PYTHON))
+ifndef PYTHON
+$(error "python3.5 is required.")
+endif
+
+# check "iota" module in Python installation
+PY_CHECK_MOD_IOTA := $(shell $(PYTHON) -c "import iota" 2>/dev/null && \
+                       echo 1 || echo 0)
+ifneq ("$(PY_CHECK_MOD_IOTA)","1")
+    $(error "dependency error $@ because PyOTA is not installed, to install the latest version: pip3 install pyota")
```
Comment (jserv) [inline mk/python.mk]: If Python executable is restricted, the validation of `pip` would be considered.

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (jserv) [reply README.md]: The explanation should be enforced to validation logic.

File: README.md [hunk 9]
```diff
@@ -10,7 +10,7 @@ Most use cases for micropayments involve a single user or device interacting
 repeatedly with a few vendors.
 
 ## Prerequisites
-`iota-swarm-node`, writing with Python3+ compatible code as well as depending on [Pyota](https://github.com/iotaledger/iota.lib.py) is theoretically available for bot Python 3.5 and 3.6; yet, our testing result indicates that the environment for Python 3.6 is invalid in this case. Details as [Pyota issue#203](https://github.com/iotaledger/iota.lib.py/issues/203).
+`iota-swarm-node`, compatible with Python3+ as well as depending on [Pyota](https://github.com/iotaledger/iota.lib.py) is theoretically available for both Python 3.5 and 3.6; yet, our testing result indicates that the environment for Python 3.6 is invalid in this case. Details as [Pyota issue#203](https://github.com/iotaledger/iota.lib.py/issues/203).
```
Comment (jserv) [inline README.md]: It is `PyOTA` instead of `Pyota`.

File: mk/python.mk [hunk 10]
```diff
@@ -1,12 +1,35 @@
...
+# check Python version
+PYTHON := $(shell which "python")
+PYTHON_COMPATIBLE = python3.5
+
 ifndef PYTHON
-$(error "python3.5 is required.")
+    $(error "Python is not executable, try to check the PATH environment variable or install $(PYTHON_COMPATIBLE) package.")
+endif # PYTHON
+
+ifdef PYTHON
```
Comment (jserv) [inline mk/python.mk]: You don't have to check `$(PYTHON)` since you already exclude its absence.

File: mk/python.mk [hunk 11]
```diff
@@ -1,12 +1,35 @@
...
+    PY_CHECK_VERSION := $(shell python -c "print(__import__('sys').version_info[0:2])")
+
+ifneq ($(PY_CHECK_VERSION), (3, 5))
+    PYTHON_COMPATIBLE_PATH := $(shell which $(PYTHON_COMPATIBLE))
+
+    ifdef PYTHON_COMPATIBLE_PATH
+        $(error "Found $(PYTHON_COMPATIBLE) in current environment, but it should be setting as \
+            default interpreter, try to create a symbolic link for $(PYTHON_COMPATIBLE_PATH)")
```
Comment (jserv) [inline mk/python.mk]: Don't suggest to make symbolic link. Instead, configure via environment variables.

File: mk/python.mk [hunk 12]
```diff
@@ -1,12 +1,35 @@
...
+    else
+        $(error "$(shell python -V):Invalid Python version, only available for $(PYTHON_COMPATIBLE).")
+    endif
 endif
+endif # PYTHON
+
+ifndef PYTHON
```
Comment (jserv) [inline mk/python.mk]: Ditto. Focus on detailed items.

File: mk/python.mk [hunk 13]
```diff
@@ -1,12 +1,35 @@
...
+endif # PYTHON
 
 # check "iota" module in Python installation
+ifdef PYTHON
 PY_CHECK_MOD_IOTA := $(shell $(PYTHON) -c "import iota" 2>/dev/null && \
                        echo 1 || echo 0)
 ifneq ("$(PY_CHECK_MOD_IOTA)","1")
     $(error "dependency error $@ because PyOTA is not installed, to install the latest version: pip3 install pyota")
 endif
+endif # PYTHON
```
Comment (jserv) [inline mk/python.mk]: Avoid nested `ifdef` and `endif` pair as possible as you can.

File: mk/python.mk [hunk 14]
```diff
@@ -1,12 +1,35 @@
...
+    PY_CHECK_VERSION := $(shell python -c "print(__import__('sys').version_info[0:2])")
+
+ifneq ($(PY_CHECK_VERSION), (3, 5))
+    PYTHON_COMPATIBLE_PATH := $(shell which $(PYTHON_COMPATIBLE))
+
+    ifdef PYTHON_COMPATIBLE_PATH
+        $(error "Found $(PYTHON_COMPATIBLE) in current environment, but it should be setting as \
+            default interpreter, try to create a symbolic link for $(PYTHON_COMPATIBLE_PATH)")
+    else
+        $(error "$(shell python -V):Invalid Python version, only available for $(PYTHON_COMPATIBLE).")
```
Comment (jserv) [inline mk/python.mk]: It is NOT `Invalid Python version`. Instead, only Python 3.5 is known to work with PyOTA. You have to address.

File: mk/python.mk [hunk 15]
```diff
@@ -1,12 +1,35 @@
-PYTHON = python3.5
-PYTHON := $(shell which $(PYTHON))
+# check Python version
+PYTHON := $(shell which "python")
+PYTHON_COMPATIBLE = python3.5
```
Comment (jserv) [inline mk/python.mk]: Rename it to `PYTHON35` to reflect its functionality.

File: README.md [hunk 16]
```diff
@@ -10,6 +10,7 @@ Most use cases for micropayments involve a single user or device interacting
 repeatedly with a few vendors.
 
 ## Prerequisites
+`iota-swarm-node`, compatible with Python3+ as well as depending on [Pyota](https://github.com/iotaledger/iota.lib.py) is theoretically available for both Python 3.5 and 3.6; yet, our testing result indicates that the environment for Python 3.6 is invalid in this case. Details as [Pyota issue#203](https://github.com/iotaledger/iota.lib.py/issues/203).
```
Comment (jserv) [inline README.md]: Correct the wording: "invalid"

File: mk/python.mk [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+ifndef PYTHON
+    $(error "Python is not executable, try to check the PATH environment variable or \
+        install python$(PYTHON_COMPATIBLE_VERSION) package.")
+endif # $(shell which "python")
+
+PY_CHECK_VERSION := $(shell python -c "print(str(__import__('sys').version_info.major) + \
+    str('.') + str(__import__('sys').version_info.minor) )")
+
+ifneq ($(PY_CHECK_VERSION), $(PYTHON_COMPATIBLE_VERSION))
+
```
Comment (jserv) [inline mk/python.mk]: Avoid this blank line.

File: mk/python.mk [hunk 18]
```diff
@@ -0,0 +1,30 @@
+# check Python version
+PYTHON := $(shell which "python")
+PYTHON_COMPATIBLE_VERSION = 3.5
+
+ifndef PYTHON
```
Comment (jserv) [inline mk/python.mk]: If `python` is not present, try `python3`. Then, if version is mismatching, try `python3.5`.

File: mk/python.mk [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+
+ifneq ($(PY_CHECK_VERSION), $(PYTHON_COMPATIBLE_VERSION))
+
+    PYTHON_COMPATIBLE_PATH := $(shell which "python"$(PYTHON_COMPATIBLE_VERSION))
+
+    ifdef PYTHON_COMPATIBLE_PATH
+        $(error "Found python$(PYTHON_COMPATIBLE_VERSION) in current environment, but it should be setting as \
+            default interpreter, try to set $(PYTHON_COMPATIBLE_PATH) in PATH environment variable.")
+    else
+        $(error "$(shell python -V):Unsupported python version, only available for Python$(PYTHON_COMPATIBLE_VERSION).")
```
Comment (jserv) [inline mk/python.mk]: Replace `python` with `$(PYTHON)`.

File: README.md (hunk 16 above)
Comment (jserv) [reply README.md]: Remove "theoretically". Instead, it would be "expectedly" or something alike.

Comment (jserv) [reply README.md]: Replace `Pyota issue#203` with `PyOTA issue #203`. Be aware of the spelling.

**Comment by yillkid** [comment]
> Use macros in GNU macro to simplify detection flow.
Could you explain more about that?

**Comment by yillkid** [comment]
> Can you handle the case that both python3.5 and python3.6 are available in $PATH?

The $(PYTHON) assign priority is python --> python3 --> python(favored version), if both python3.5, python3.6 are available in $PATH and symbolic link `python` is linked with python3.5, it will work. If not, an error message `Found python$(PY_FAVORED_VER), but it should be set as default interpreter, try to set $(PYTHON) in PATH environment variable.` will be trigged.

File: README.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (jserv) [reply README.md]: Rewrite the abobe wording to
> "This package depends in PyOTA, which is known to work with Python 3.5. However, before PyOTA issue #203 is resolved, Python 3.6/3.7 will be excluded for running this package. The build system would detect and validate the environment in advance."

File: mk/python.mk [hunk 20]
```diff
@@ -0,0 +1,44 @@
+# check Python version
+PY_FAVORED_VER = 3.5
+PYTHON := $(shell which "python")
+ifndef PYTHON
+    PYTHON := $(shell which "python3")
```
Comment (jserv) [inline mk/python.mk]: Don't indent here.

File: mk/python.mk [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+endif
+
+ifndef PYTHON
+    PYTHON := $(shell which "python$(PY_FAVORED_VER)")
```
Comment (jserv) [inline mk/python.mk]: Ditto.

File: mk/python.mk [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+endif
+
+ifndef PYTHON
+    $(error "Python is not executable, try to check the PATH environment variable or \
```
Comment (jserv) [inline mk/python.mk]: Ditto.

File: mk/python.mk [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+        install python$(PY_FAVORED_VER) package.")
+endif # $(shell which "python")
+
+PY_CHECK_VERSION := $(shell $(PYTHON) -V)
+
+ifeq "$(filter $(PY_FAVORED_VER).%, $(PY_CHECK_VERSION))" ""
+    PYTHON := $(shell which "python"$(PY_FAVORED_VER))
```
Comment (jserv) [inline mk/python.mk]: Reduce the indentation for pretty code snip.

File: mk/python.mk [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+    else
+        $(error "$(shell python -V):Unsupported python version, only available for Python$(PY_FAVORED_VER).")
+    endif
+endif
+
+# check "iota" module in Python installation
+PY_CHECK_MOD_IOTA := $(shell python -c "import iota" 2>/dev/null && \
+                       echo 1 || echo 0)
+ifneq ("$(PY_CHECK_MOD_IOTA)","1")
+    PIP := $(shell pip -V)
```
Comment (jserv) [inline mk/python.mk]: Ditto.

File: mk/python.mk [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+    PIP_CHECK_VERSION := $(findstring python 3, $(PIP))
+
+ifndef PIP
+    $(error "PIP package doesn't pre-configured, try to install it.")
+endif
+
+ifndef PIP_CHECK_VERSION
+    PIP_FAVORED_VER = 3
+endif
+    $(error "PyOTA package doesn't pre-configured, install the latest version: pip$(PIP_FAVORED_VER) install pyota")
```
Comment (jserv) [inline mk/python.mk]: Ditto.

**Comment by jserv** [comment]
Check dcurl for illustrating macro usage with GNU make: https://github.com/DLTcollab/dcurl/blob/dev/Makefile#L32

File: README.md [hunk 26]
```diff
@@ -10,7 +10,7 @@ Most use cases for micropayments involve a single user or device interacting
 repeatedly with a few vendors.
 
 ## Prerequisites
-`iota-swarm-node`, compatible with Python3+ as well as depending on [Pyota](https://github.com/iotaledger/iota.lib.py) is theoretically available for both Python 3.5 and 3.6; yet, our testing result indicates that the environment for Python 3.6 is invalid in this case. Details as [Pyota issue#203](https://github.com/iotaledger/iota.lib.py/issues/203).
+This package depends in PyOTA, which is known to work with Python 3.5. However, before PyOTA issue [#203](https://github.com/iotaledger/iota.lib.py/issues/203) is resolved, Python 3.6/3.7 will be excluded for running this package. The build system would detect and validate the environment in advance.
```
Comment (jserv) [inline README.md]: depends "on", not "in".

File: mk/python.mk [hunk 27]
```diff
@@ -2,43 +2,46 @@
...
-    ifdef PYTHON
-        $(error "Found python$(PY_FAVORED_VER), but it should be set as \
-            default interpreter, try to set $(PYTHON) in PATH environment variable.")
-    else
-        $(error "$(shell python -V):Unsupported python version, only available for Python$(PY_FAVORED_VER).")
-    endif
-endif
+PYTHON := $(shell which "python"$(PY_FAVORED_VER))
+ifdef PYTHON
+$(error "Found python$(PY_FAVORED_VER), but it should be set as \
```
Comment (jserv) [inline mk/python.mk]: Do you think it is valid according to the error message?

File: mk/python.mk [hunk 28]
```diff
@@ -2,43 +2,46 @@
...
                        echo 1 || echo 0)
-ifneq ("$(PY_CHECK_MOD_IOTA)","1")
-    PIP := $(shell pip -V)
-    PIP_CHECK_VERSION := $(findstring python 3, $(PIP))
+ifeq ("$(PY_CHECK_MOD_IOTA)","0")
+PIP := $(shell pip -V && echo 1 || echo 0))
+PIP_CHECK_VERSION := $(findstring python 3, $(PIP) && echo $@ || echo 0)
+endif
+
+ifneq ("$(PIP)", "1")
```
Comment (jserv) [inline mk/python.mk]: Use `ifdef` rather than `ifneq` or `ifeq` for consistency.

File: mk/python.mk [hunk 29]
```diff
@@ -2,43 +2,46 @@
...
+$(error "PIP package doesn't pre-configured, try to install it.")
```
Comment (jserv) [inline mk/python.mk]: Fix grammar error.

File: mk/python.mk [hunk 30]
```diff
@@ -2,43 +2,46 @@
...
-ifndef PIP
-    $(error "PIP package doesn't pre-configured, try to install it.")
+ifeq ("$(PIP_CHECK_VERSION)", "0")
+PIP_FAVORED_VER = 3
 endif
 
-ifndef PIP_CHECK_VERSION
-    PIP_FAVORED_VER = 3
+ifeq ("$(PY_CHECK_MOD_IOTA)", "0")
+$(error "PyOTA package doesn't pre-configured, install the latest version: pip$(PIP_FAVORED_VER) install pyota")
```
Comment (jserv) [inline mk/python.mk]: Ditto. Fix grammar.

Comment (jserv) [reply mk/python.mk]: Using `pip` is not the only way to install Python packages. Sometimes, package maintainer might prepare prebuilt one or recipes that describe how to install Python packages. You can leave the message: "please ask your administrator to install the package 'pyota' or build from source: https://github.com/iotaledger/iota.lib.py"

File: mk/python.mk [hunk 31]
```diff
@@ -2,43 +2,46 @@
...
 
 ifndef PYTHON
-    PYTHON := $(shell which "python$(PY_FAVORED_VER)")
+PYTHON := $(shell which "python$(PY_FAVORED_VER)")
 endif
 
 ifndef PYTHON
-    $(error "Python is not executable, try to check the PATH environment variable or \
-        install python$(PY_FAVORED_VER) package.")
+$(error "Python is not executable, try to check the PATH environment variable or \
```
Comment (jserv) [inline mk/python.mk]: It is ambiguous to say "Python is not executable". Use precise words.

File: mk/python.mk [hunk 32]
```diff
@@ -0,0 +1,47 @@
+# check Python version
+PY_FAVORED_VER = 3.5
+PYTHON := $(shell which "python")
+ifndef PYTHON
+PYTHON := $(shell which "python3")
+endif
+
```
Comment (jserv) [inline mk/python.mk]: No blank line here.

File: mk/python.mk [hunk 33]
```diff
@@ -0,0 +1,47 @@
...
+ifndef PYTHON
+PYTHON := $(shell which "python$(PY_FAVORED_VER)")
+endif
+
+ifndef PYTHON
+$(error "Python is not executable, try to check the PATH environment variable or \
+install python$(PY_FAVORED_VER) package.")
+endif # $(shell which "python")
```
Comment (jserv) [inline mk/python.mk]: Remove non-effective "$(shell which "python")".

Comment (jserv) [reply mk/python.mk]: Let's say, "no valid Python interpreter is found".

File: mk/python.mk [hunk 34]
```diff
@@ -0,0 +1,47 @@
...
+
+PY_CHECK_VERSION := $(shell $(PYTHON) -V)
+
+ifeq "$(filter $(PY_FAVORED_VER).%, $(PY_CHECK_VERSION))" ""
+PYTHON := $(shell which "python"$(PY_FAVORED_VER))
+ifdef PYTHON
+$(error "Found python$(PY_FAVORED_VER), but it should be set as \
+default interpreter, try to set $(PYTHON) in PATH environment variable.")
+else
+$(error "$(shell python -V):Unsupported python version, only available for Python$(PY_FAVORED_VER).")
```
Comment (jserv) [inline mk/python.mk]: Explain the exact reason and point unresolved issue(s) here.

File: mk/python.mk [hunk 35]
```diff
@@ -0,0 +1,47 @@
...
+endif # PYTHON
+endif # "$(filter ...
```
Comment (jserv) [inline mk/python.mk]: Ditto. You don't have to mention "$(filter ..." once you remove nested conditions.

File: mk/python.mk (hunk 34 above)
Comment (jserv) [reply mk/python.mk]: Validate ifdef-else statement. Logic error.

**Comment by jserv** [comment]
@yillkid, you should create a new branch as workspace for commit "Store transaction before broadcast
". Otherwise, this pull request would be polluted.

File: README.md [hunk 36]
```diff
@@ -10,6 +10,7 @@ Most use cases for micropayments involve a single user or device interacting
 repeatedly with a few vendors.
 
 ## Prerequisites
+This package depends on PyOTA, which is known to work with Python 3.5. However, before PyOTA issue [#203](https://github.com/iotaledger/iota.lib.py/issues/203) is resolved, Python 3.6/3.7 will be excluded for running this package. The build system would detect and validate the environment in advance.
```
Comment (wusyong) [inline README.md]: PyOTA doesn't support Python 3.7 because it has package `Filters` which doesn't either.
